libgalois/include/katana/analytics/ClusteringImplementationBase.h

```diff
@@ -270,53 +266,57 @@ struct ClusteringImplementationBase {
     return max_index;
   }
 
-  /**
-   * Computes the modularity gain of the current cluster assignment.
-   */
-  template <typename EdgeWeightType>
-  static double CalModularity(
-      const Graph& graph, CommunityArray& c_info, double& e_xx, double& a2_x,
-      double& constant_for_second_term) {
-    /* Variables needed for Modularity calculation */
-    double mod = -1;
-
+  template <
+      typename EdgeWeightType, typename CommunityIDType,
+      typename NodeWeightFunc>
+  static double ModularityImpl(
+      const Graph& graph, const NodeWeightFunc& node_wt_func, double& e_xx,
+      double& a2_x, const double constant_for_second_term) {
     katana::NUMAArray<EdgeTy> cluster_wt_internal;
-
-    /*** Initialization ***/
     cluster_wt_internal.allocateBlocked(graph.num_nodes());
+    katana::ParallelSTL::fill(
+        cluster_wt_internal.begin(), cluster_wt_internal.end(), 0);
 
     /* Calculate the overall modularity */
     katana::GAccumulator<double> acc_e_xx;
     katana::GAccumulator<double> acc_a2_x;
 
-    katana::do_all(
-        katana::iterate(graph), [&](GNode n) { cluster_wt_internal[n] = 0; });
-
     katana::do_all(katana::iterate(graph), [&](GNode n) {
-      auto n_data_current_comm_id =
-          graph.template GetData<CurrentCommunityID>(n);
-      for (auto ii = graph.edge_begin(n); ii != graph.edge_end(n); ++ii) {
-        if (graph.template GetData<CurrentCommunityID>(graph.GetEdgeDest(ii)) ==
-            n_data_current_comm_id) {
+      auto n_data_current_comm = graph.template GetData<CommunityIDType>(n);
+      for (auto e : graph.edges(n)) {
+        if (graph.template GetData<CommunityIDType>(graph.edge_dest(e)) ==
+            n_data_current_comm) {
           cluster_wt_internal[n] +=
-              graph.template GetEdgeData<EdgeWeight<EdgeWeightType>>(ii);
+              graph.template GetEdgeData<EdgeWeight<EdgeWeightType>>(e);
         }
       }
     });
 
     katana::do_all(katana::iterate(graph), [&](GNode n) {
       acc_e_xx += cluster_wt_internal[n];
-      acc_a2_x +=
-          (double)(c_info[n].degree_wt) *
-          ((double)(c_info[n].degree_wt) * (double)constant_for_second_term);
+      double degree_wt = node_wt_func(n);
```

Comment:
this is problematic. should instead be `community_wt_func`

Reply:
Are you saying name needs to change?

Reply:
yeah, that's what I meant. But no problem, we can rename it later on whenever we make our next update (which is going to be soon to fix leiden anyways). thanks